src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -265,15 +277,56 @@ public NodeStateTable<T> ToImmutableAndFree()
                     return NodeStateTable<T>.Empty;
                 }
 
-                var hasNonCached = _states.Any(static s => s.Any(static i => i.Item2 != EntryState.Cached));
+                var hasNonCached = _states.Any(static s => !s.IsCached);
                 return new NodeStateTable<T>(_states.ToImmutableAndFree(), isCompacted: !hasNonCached);
-
             }
 
             internal ImmutableArray<T> GetLastEntries()
             {
-                return _states[_states.Count - 1].SelectAsArray(t => t.Item1);
+                var nodeStateTableEntry = _states[_states.Count - 1];
+                return nodeStateTableEntry.Item is object ? ImmutableArray.Create<T>(nodeStateTableEntry.Item) : nodeStateTableEntry.Items;
+            }
+        }
+
+        private readonly struct TableEntry
+        {
+            internal readonly ImmutableArray<T> Items;
+
+            internal readonly T? Item;
+
+            internal readonly EntryState? State;
+
+            internal readonly ImmutableArray<EntryState> States;
+
+            public TableEntry(T item, EntryState state)
+                : this(item, default, state, default) { }
+
+            public TableEntry(ImmutableArray<T> items, EntryState state)
+                : this(default, items, state, default) { }
+
+            public TableEntry(ImmutableArray<T> items, ImmutableArray<EntryState> states)
+                : this(default, items, null, states) { }
+
+            private TableEntry(T? item, ImmutableArray<T> items, EntryState? state, ImmutableArray<EntryState> states)
+            {
+                Debug.Assert(item is object || !items.IsDefault);
```

Comment:
>|| [](http://example.com/codeflow?start=44&length=2)

`==`?

Reply:
@cston / @jaredpar  Hmm, this is a good point. The assertion doesn't hold because we allow value types, so its technically valid for both parameters to be set in that case (albeit with item being the default value). 

That got me thinking though, and i've realized we're making an assumption that doesn't hold because there is nothing stopping users from returning a null value in the higher level transform, meaning its valid to have a `null` as an item state.

We could require non-null values in the transforms? I'm not sure if we want to make that restriction however. For now I'm just going to remove this assertion altogether and have added a test confirming we can put null values in the table.